### PR TITLE
Change default of antialias parameter from None to 'warn'

### PR DESCRIPTION
**Pull request by NicolasHug**
Before changing the default to True in 2 versions

Closes https://github.com/pytorch/vision/issues/7093

### CODE REVIEW

File: torchvision/transforms/functional.py [hunk 1]
```diff
@@ -1578,3 +1608,21 @@ def elastic_transform(
     if not isinstance(img, torch.Tensor):
         output = to_pil_image(output, mode=img.mode)
     return output
+
+
+# TODO in v0.17: remove this helper and change default of antialias to True everywhere
+def _check_antialias(antialias: Optional[Union[str, bool]], interpolation: InterpolationMode) -> Optional[bool]:
+    if isinstance(antialias, str):  # it should be "warn", but we don't bother checking against that
+        # TODO Should we only warn if input is a tensor, or for PIL images too?
```
Comment (NicolasHug) [inline torchvision/transforms/functional.py:1616]: Thoughts?

File: torchvision/transforms/functional_tensor.py [hunk 2]
```diff
@@ -451,7 +453,11 @@ def resize(
         antialias = False
 
     if antialias and interpolation not in ["bilinear", "bicubic"]:
-        raise ValueError("Antialias option is supported for bilinear and bicubic interpolation modes only")
+        # We manually set it to False to avoid an error downstream in interpolate()
+        # This behaviour is documented: the parameter is irrelevant for modes
+        # that are not bilinear or bicubic. We used to raise an error here, but
+        # now we don't as True is the default.
```
Comment (NicolasHug) [inline torchvision/transforms/functional_tensor.py:459]: We used to raise an error, now we silently disable antialias. Otherwise, we'd be breaking those use-cases when we switch the default to True in 2 versions:

```py
Resize(interpolation=NEAREST)(img)  # uh oh, default changed to True, which raises an error in main branch rn.
```

This is a slight change of behaviour and it's not perfect, but I feel like the alternative is worse: we could introduce another value for `antialias` e.g. `"auto"`, which would mean "Do what PIL does i.e. apply antialias for bilinear and bicubic, and no antialias otherwise". But having that + True + False + None will bloat the API IMHO.

If it makes it any better, remember that PIL already silently switches antialiasing on or off depending on the interpolation mode (granted, they don't have an `antialias` parameter so it's not a completely fair comparison).

**Review by NicolasHug — COMMENTED** [review]
@pmeier @vfdev-5 this is still WIP but I'm curious about your initial thoughts on the points noted above

File: torchvision/transforms/functional.py (hunk 1 above)
Comment (pmeier) [reply torchvision/transforms/functional.py:1616]: I would only warn for tensors. Since the value is, and will be, ignored for PIL and we already warn that it is passed in the first place 

https://github.com/pytorch/vision/blob/135a0f9ea9841b6324b4fe8974e2543cbb95709a/torchvision/transforms/functional.py#L474-L476

the warning here is irrelevant for the user.

Comment (NicolasHug) [reply torchvision/transforms/functional.py]: OK, we're only warning for Tensors now. One minor caveat is that since we can't warn in `__init__()` of `Resize()` anymore, the warning will only happen during `forward()`. I assume that's fine? Right now, I tend to prefer that to the alternative of warning for PIL users who should not have to worry about all this

File: torchvision/prototype/datapoints/_bounding_box.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def resize(  # type: ignore[override]
         size: List[int],
         interpolation: InterpolationMode = InterpolationMode.BILINEAR,
         max_size: Optional[int] = None,
-        antialias: Optional[bool] = None,
+        antialias: Optional[Union[str, bool]] = "warn",
```
Comment (vfdev-5) [inline torchvision/prototype/datapoints/_bounding_box.py:81]: Technically, it is not Optional anymore as default is "warn" and not None
```suggestion
        antialias: Union[str, bool] = "warn",
```

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -308,13 +308,27 @@ class Resize(torch.nn.Module):
             smaller edge may be shorter than ``size``. This is only supported
             if ``size`` is an int (or a sequence of length 1 in torchscript
             mode).
-        antialias (bool, optional): antialias flag. If ``img`` is PIL Image, the flag is ignored and anti-alias
-            is always used. If ``img`` is Tensor, the flag is False by default and can be set to True for
-            ``InterpolationMode.BILINEAR`` and ``InterpolationMode.BICUBIC`` modes.
-            This can help making the output for PIL images and tensors closer.
+        antialias (bool, optional): Whether to apply antialiasing.
```
Comment (NicolasHug) [inline torchvision/transforms/transforms.py:311]: Updated the docstring hopefully making it more obvious. You're allowed to request changes but only once :) !
(I want to avoid having to update those in multiple places 10x in a row )

File: torchvision/prototype/datapoints/_bounding_box.py (hunk 3 above)
Comment (NicolasHug) [reply torchvision/prototype/datapoints/_bounding_box.py:81]: Honestly, I have no clue what's the difference between `Optional[bool]` and `Union[NoneType, bool]`. @pmeier any thoughts?

Comment (vfdev-5) [reply torchvision/prototype/datapoints/_bounding_box.py:81]: I think, `Optional[X]` is `Union[X, None]`. Anyway, it is a nit, we can ignore that.

File: torchvision/transforms/transforms.py (hunk 4 above)
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:311]: I was wondering if we could later also opt to the similar idea as pytorch is doing for their docs: 

```python

geom_common_args = {
    "antialias": "<doctring>"
}

class ATranform(...)
"""Doc about ATrasform

    {antialias}
""".format(**geom_common_args)
```


https://github.com/pytorch/pytorch/blob/98e1b3e93a10b7dceeea7df489dafacbcaab556f/torch/_torch_docs.py#L190-L214

Comment (NicolasHug) [reply torchvision/transforms/transforms.py:311]: So far we've found value in having the docstings easily available when reading the code. It's true that this kind of docstring engine becomes necessary when the same parameters are needed across many different APIs. We might have reached that point already, IDK. Happy to discuss more about this later

File: test/test_prototype_transforms.py [hunk 5]
```diff
@@ -1885,3 +1886,67 @@ def test__transform(self, inpt):
         assert type(output) is type(inpt)
         assert output.shape[-4] == num_samples
         assert output.dtype == inpt.dtype
+
+
+# TODO: remove this test in 0.17 when the default of antialias changes to True
+def test_antialias_warning():
```
Comment (NicolasHug) [inline test/test_prototype_transforms.py:1892]: I'm mixing the functional and class tests here... Hope that's fine (I found it to be a awkward to split them as done in the V1 tests)

**Comment by NicolasHug** [comment]
I just confirmed on the references that the behaviour looks fine. I'm getting warnings when relevant (only on tensors) and the accuracy boost is consistent with https://github.com/pytorch/vision/issues/6506 (I only checked `mobilenet_v3_large` this time). On the segmentation references, there's no perf difference on the model I tried (FCN resnet 50). On the other references like segmentation or video, the resizing was done directly on tensors (in the models or in the presets), so there's no change there.

File: test/test_prototype_transforms.py (hunk 5 above)
Comment (pmeier) [reply test/test_prototype_transforms.py:1892]: Fine by me. Test is self-contained enough that there shouldn't be any confusion about it later.

File: test/test_functional_tensor.py [hunk 6]
```diff
@@ -531,8 +532,8 @@ def test_resize(device, dt, size, max_size, interpolation):
         tensor = tensor.to(dt)
         batch_tensors = batch_tensors.to(dt)
 
-    resized_tensor = F.resize(tensor, size=size, interpolation=interpolation, max_size=max_size)
-    resized_pil_img = F.resize(pil_img, size=size, interpolation=interpolation, max_size=max_size)
+    resized_tensor = F.resize(tensor, size=size, interpolation=interpolation, max_size=max_size, antialias=True)
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_functional_tensor.py:535]: I don't think v2 tests need changes at all:

1. The tensor kernel, e.g. `resize_image_tensor` only branches on `antialias` to bail out in case it is set together with an `interpolation` that doesn't support it: 

   https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/torchvision/prototype/transforms/functional/_geometry.py#L150-L153

   Thus, our `sample_inputs_fn`'s, that should cover all _valid_ branches, don't set it at all:

   - https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/prototype_transforms_kernel_infos.py#L265
   - https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/prototype_transforms_kernel_infos.py#L275
   - https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/prototype_transforms_kernel_infos.py#L277
2. Our reference test check the image tensor against the corresponding PIL kernel, e.g. `resize_image_pil`. Since PIL "hardcodes" `antialias=True`, we need to do the same:

   https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/prototype_transforms_kernel_infos.py#L280-L286

   https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/prototype_transforms_kernel_infos.py#L304-L309

   (I saw that I messed up the `antialias` value here. Will send a PR soon)
3. Since the dispatcher, e.g. `resize`, no longer contains any logic besides the dispatching, we just use the kernel inputs and check if they are dispatched to the correct kernel:

   https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/prototype_transforms_dispatcher_infos.py#L158-L170

   - https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/test_prototype_transforms_functional.py#L358
   - https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/test_prototype_transforms_functional.py#L383
   - https://github.com/pytorch/vision/blob/2cd25c1a05012a3720a87f20cec436811fadeedd/test/test_prototype_transforms_functional.py#L418

File: torchvision/transforms/functional.py [hunk 7]
```diff
@@ -431,10 +431,24 @@ def resize(
...
+              Other mode aren't affected. This is probably what you want to use.
+            - ``False``: will not apply antialiasing for tensors on any mode. PIL
+              images are still antialiased on bilinear or bicubic modes, because
+              PIL doesn't support no antialias.
+            - ``None``: equivalent to ``False`` for tensors and ``True`` for
+              PIL images. This value exists for legacy reasons and you probably
+              don't want to use it unless you really know what you are doing.
+
+            The current default is ``None`` **but will change to** ``True`` **in
+            v0.17** for the PIL and Tensor backends to be consistent.
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/transforms/functional.py:451]: None can still be useful because `False` isn't supported by PIL images

https://github.com/pytorch/vision/issues/7093

> Side note: It is worth keeping the None option: since PIL doesn't support no-antialiasing, antialias=False would raise an annoying warning when using it with PIL. This could typically be needed in training pipelines where multiple antialias strategies or backends are in use (something we may eventually do in torchvision....)

File: torchvision/transforms/functional.py [hunk 8]
```diff
@@ -471,6 +485,8 @@ def resize(
     if (image_height, image_width) == output_size:
         return img
 
+    antialias = _check_antialias(img, antialias, interpolation)
+
     if not isinstance(img, torch.Tensor):
         if antialias is not None and not antialias:
             warnings.warn("Anti-alias option is always applied for PIL Image input. Argument antialias is ignored.")
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/transforms/functional.py:492]: Yes, `is False` is clearer, althought I avoided doing this in this PR because it's not strictly needed for the original purpose, and the PR is already pretty long. I don't mind doing if you think it's less confusing

File: test/test_functional_tensor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/test_functional_tensor.py:535]: We're still getting hundreds of warnings just in `test_prototype_transforms.py`. Do you know how to cleanly get rid of those?

```
test/test_prototype_transforms.py: 456 warnings
  /home/nicolashug/dev/vision/torchvision/transforms/functional.py:1621: UserWarning: The default value of the antialias parameter will change from None to True in v0.17, in order to be consistent across the PIL and Tensor backends. To suppress this warning, directly pass antialias=True (recommended, future default), antialias=None (current default, which means False for Tensors and True for PIL), or antialias=False (only works on Tensors - PIL will still use antialiasing).
```

File: torchvision/transforms/functional.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/transforms/functional.py:492]: I don't care if we do it here or in a follow-up, but we should do it. I had to build the table above to see if the new case of `"warn"` does something unexpected and that is not a good sign for an otherwise simple check.

File: test/test_functional_tensor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_functional_tensor.py:535]: Fixed in f6c8140cad81303747dcd5b1ff3b61e79af1ea4a. It was just two transforms that needed fixing. They generated ~400 warnings since we tested a lot of parametrizations on them.

File: torchvision/transforms/functional.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply torchvision/transforms/functional.py:483]: @pmeier `antialias="warn"` should behave as `antialias=None`, so, I would say there is no 4th line in the table (also values of the 4th line seem to be incorrect).

File: test/test_functional_tensor.py [hunk 9]
```diff
@@ -483,8 +484,8 @@ def test_resize(device, dt, size, max_size, interpolation):
...
+    resized_pil_img = F.resize(pil_img, size=size, interpolation=interpolation, max_size=max_size, antialias=True)
```
Comment (vfdev-5) [inline test/test_functional_tensor.py:488]: Nit: As we now doing `antialias=True`, we can also reduce the `tol` below now:
```
_assert_approx_equal_tensor_to_pil(resized_tensor_f, resized_pil_img, tol=8.0)
```

Comment (NicolasHug) [reply test/test_functional_tensor.py:488]: OK, I'll merge now to unblock, and I'll open a PR to check the CI on a lower tol

File: torchvision/transforms/functional.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/transforms/functional.py:483]: Well, the result is the same, but the way it is achieved is different. What do you think is wrong?

```pycon
>>> antialias = "warn"
>>> antialias is not None
True
>>> not antialias
False
>>> antialias is not None and not antialias
False
```
